Packs/Claroty/Integrations/Claroty/Claroty.py

```diff
@@ -0,0 +1,627 @@
+    "High": 3,
+    "Critical": 4
+}
+ALERT_CTD_FIELD_TO_DEMISTO_FIELD = {
+    'resource_id': "ResourceID",
+    'type': "AlertTypeID",
+    'type__': "AlertType",
+    'severity__': "Severity",
+    'network_id': "NetworkID",
+    'resolved': "Resolved",
+    'description': "Description",
+    'actionable_assets': "RelatedAssets",
+    'alert_indicators': "Indicator",
+    'category__': "Category",
+    'timestamp': "Timestamp"
+}
+ASSET_CTD_FIELD_TO_DEMISTO_FIELD = {
+    'id': "AssetID",
+    'name': "Name",
+    'insight_names': "InsightName",
+    'vendor': "Vendor",
+    "criticality__": "Criticality",
+    'asset_type__': "AssetType",
+    "last_seen": "LastSeen",
+    "ipv4": "IP",
+    "mac": "MAC",
+    'virtual_zone_name': "VirtualZone",
+    "class_type": "ClassType",
+    'site_name': "SiteName",
+    "project_parsed": "WasParsed",
+    "risk_level": "RiskLevel",
+    "firmware": "FirmwareVersion",
+    "resource_id": "ResourceID",
+    "site_id": "SiteID",
+    "insights": "Insights"
+}
+RESOLVE_STRING_TO_TYPE = {
+    "resolve": 1,
+    "archive": 2
+}
+DEFAULT_ALERT_FIELD_LIST = ["resource_id", "type", "severity", "network_id", "resolved", "description",
+                            "alert_indicators", "actionable_assets", "category"]
+DEFAULT_ASSET_FIELD_LIST = ["id", "name", "insight_names", "vendor", "criticality", "asset_type", "last_seen", "ipv4",
+                            "mac", "virtual_zone_name", "class_type", "site_name", "site_id", "project_parsed",
+                            "risk_level", "firmware", "resource_id"]
+WINDOWS_CVE_BASE_URL = "ranger/insight_details/Windows%20CVEs?&format=asset_page&sort=-Score%20(CVSS)" \
+                       "&per_page=1000&page=1&id__exact="
+FULL_MATCH_BASE_URL = "ranger/insight_details/Full%20Match%20CVEs?&format=asset_page&sort=-Score%20(CVSS)" \
+                      "&per_page=1000&page=1&id__exact="
+DEFAULT_RESOLVE_ALERT_COMMENT = "Resolved by Demisto"
+MAX_ASSET_LIMIT = 5000
+MAX_ALERT_LIMIT = 8000
+
+
+class Client(BaseClient):
+    def __init__(self, **kwargs):
+        self._credentials = kwargs.pop("credentials", (None, None))
+        not_mock = kwargs.pop("not_mock", True)
+        super().__init__(**kwargs)
+        self._headers = DEFAULT_HEADERS
+        if not_mock:
+            self._generate_token()
+        else:
+            self._headers['Authorization'] = "ok"
+
+        self._list_to_filters: dict = {'alerts': [], 'assets': []}
+
+    def _request_with_token(self, url_suffix: str, method: str = "GET",
+                            data: dict = None) -> Union[Dict, str, requests.Response]:
+        try:
+            return self._http_request(method, url_suffix=url_suffix, data=data)
+        except DemistoException:
+            demisto.setIntegrationContext({'jwt_token': None})
+            self._headers.pop('Authorization', None)
+            # assuming it was just the token that expired, retrying to send the request with the new token
```

Comment:
We are returning only the JWT and not the expiration date from our API (the whole thing is kind of encapsulated in our BE to keep it secure, once a request comes and the user gets a 401, our FE will disconnect and will force the user to login again).
We thought that adding a timer that is kept only on the demisto app is a bit weird (what happens if the default timeout is changed in CTD?)... So we're just retrying the request with a new token - if it succeeded all is well and we can continue in the flow, otherwise we'll get permission denied again and stop asking for new tokens

Reply:
the JWT itself actually contains the expiration (try to paste it in https://jwt.io and see the "exp" header) . To extract it you could use the PyJWT library. I can show you how to do it if you want. We do have a python image that already contains pyjwt.

Reply:
We can't since we don't return the expiry